File: getting-started/variant-ts/module-09-capstone-wordle/exercise-01-wordle/solution/ui/terminal.ts
import { type LetterResult, type GuessResult, MAX_GUESSES, WORD_LENGTH } from "../game/game";

const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";
const DIM = "\x1b[2m";
const GREEN = "\x1b[48;5;34;97m";   // green bg, bright white
const YELLOW = "\x1b[48;5;178;97m"; // amber bg, bright white
const GRAY = "\x1b[48;5;240;97m";   // gray bg, bright white
const DIM_BG = "\x1b[48;5;236;37m"; // dark bg, dim text (empty tiles)

// Tracks the best-known state per letter across all guesses.
// correct > misplaced > absent. Unseen letters aren't in the map.
const RESULT_PRIORITY = { absent: 0, misplaced: 1, correct: 2 } as const;
const keyboard = new Map<string, LetterResult>();

const colorFor = (r: LetterResult): string => {
  switch (r) {
    case "correct": return GREEN;
    case "misplaced": return YELLOW;
    case "absent": return GRAY;
  }
};

const updateKeyboard = (result: GuessResult): void => {
  for (let i = 0; i < result.guess.length; i++) {
    const ch = result.guess[i];
    const newResult = result.letters[i];
    const prev = keyboard.get(ch);
    if (!prev || RESULT_PRIORITY[newResult] > RESULT_PRIORITY[prev]) {
      keyboard.set(ch, newResult);
    }
  }
};

// --- Tile rendering ---

const tile = (ch: string, r: LetterResult): string =>
  `${colorFor(r)} ${ch.toUpperCase()} ${RESET}`;

const emptyTile = (): string =>
  `${DIM_BG}   ${RESET}`;

const renderGuessRow = (result: GuessResult): string => {
  const tiles = Array.from(result.guess).map((ch, i) => tile(ch, result.letters[i]));
  return "    " + tiles.join(" ");
};

const renderEmptyRow = (): string => {
  const tiles = Array.from({ length: WORD_LENGTH }, () => emptyTile());
  return "    " + tiles.join(" ");
};

// --- Board ---

const printBoard = (guesses: GuessResult[]): void => {
  console.log();
  for (let i = 0; i < MAX_GUESSES; i++) {
    if (i < guesses.length) {
      console.log(renderGuessRow(guesses[i]));
    } else {
      console.log(renderEmptyRow());
    }
  }
};

// --- Keyboard ---

const KEY_ROWS = ["qwertyuiop", "asdfghjkl", "zxcvbnm"] as const;

const printKeyboard = (): void => {
  console.log();
  for (let i = 0; i < KEY_ROWS.length; i++) {
    const pad = " ".repeat((i * 2) + 4);
    const keys = Array.from(KEY_ROWS[i]).map((ch) => {
      const result = keyboard.get(ch);
      const upper = ch.toUpperCase();
      if (!result) return ` ${upper} `;
      return `${colorFor(result)} ${upper} ${RESET}`;
    });
    console.log(pad + keys.join(""));
  }
  console.log();
};

// --- Public API ---

export const displayWelcome = (): void => {
  console.log();
  console.log(`    ${BOLD} W  O  R  D  L  E ${RESET}`);
  console.log();
  console.log("    Guess the 5-letter word in 6 tries.");
  console.log();
  console.log(`    ${GREEN} E ${RESET} correct   ${YELLOW} E ${RESET} wrong spot   ${GRAY} E ${RESET} not in word`);
};

// displayTurn redraws the full board and keyboard after every guess.
export const displayTurn = (guesses: GuessResult[]): void => {
  for (const g of guesses) {
    updateKeyboard(g);
  }
  printBoard(guesses);
  printKeyboard();
};

export const readGuess = (attemptNum: number): string => {
  const raw = prompt(`    Guess ${attemptNum}/${MAX_GUESSES}:`) ?? "";
  return raw.trim().toLowerCase();
};

export const displayError = (msg: string): void => {
  console.log(`    ${DIM}-- ${msg}${RESET}\n`);
};

export const displayWin = (attempts: number): void => {
  const msgs = ["", "Genius!", "Magnificent!", "Impressive!", "Solid!", "Good job!", "Phew!"];
  console.log(`\n    ${BOLD}${msgs[attempts]}${RESET} Got it in ${attempts}.\n`);
};

export const displayLoss = (target: string): void => {
  console.log(`\n    The word was ${BOLD}${target.toUpperCase()}${RESET}. Better luck next time.\n`);
};

File: getting-started/variant-go/module-09-capstone-wordle/exercise-01-wordle/solution/ui/terminal.go
package ui

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"

	"wordle/game"
)

const (
	reset  = "\033[0m"
	bold   = "\033[1m"
	dim    = "\033[2m"
	green  = "\033[48;5;34;97m"  // green bg, bright white
	yellow = "\033[48;5;178;97m" // amber bg, bright white
	gray   = "\033[48;5;240;97m" // gray bg, bright white
	dimBg  = "\033[48;5;236;37m" // dark bg, dim text (empty tiles)
)

var reader = bufio.NewReader(os.Stdin)

// Tracks the best-known state per letter across all guesses.
// Correct > Misplaced > Absent. Unseen letters aren't in the map.
var keyboard = map[rune]int{}

func colorFor(r game.LetterResult) string {
	switch r {
	case game.Correct:
		return green
	case game.Misplaced:
		return yellow
	default:
		return gray
	}
}

func updateKeyboard(result game.GuessResult) {
	for i, ch := range result.Guess {
		v := int(result.Letters[i])
		if prev, seen := keyboard[ch]; !seen || v > prev {
			keyboard[ch] = v
		}
	}
}

// --- Tile rendering ---

func tile(ch rune, r game.LetterResult) string {
	return fmt.Sprintf("%s %c %s", colorFor(r), unicode.ToUpper(ch), reset)
}

func emptyTile() string {
	return fmt.Sprintf("%s   %s", dimBg, reset)
}

func renderGuessRow(result game.GuessResult) string {
	tiles := make([]string, game.WordLength)
	for i, ch := range result.Guess {
		tiles[i] = tile(ch, result.Letters[i])
	}
	return "    " + strings.Join(tiles, " ")
}

func renderEmptyRow() string {
	tiles := make([]string, game.WordLength)
	for i := range tiles {
		tiles[i] = emptyTile()
	}
	return "    " + strings.Join(tiles, " ")
}

// --- Board ---

func printBoard(guesses []game.GuessResult) {
	fmt.Println()
	for i := range game.MaxGuesses {
		if i < len(guesses) {
			fmt.Println(renderGuessRow(guesses[i]))
		} else {
			fmt.Println(renderEmptyRow())
		}
	}
}

// --- Keyboard ---

var keyRows = [3]string{"qwertyuiop", "asdfghjkl", "zxcvbnm"}

func printKeyboard() {
	fmt.Println()
	for i, row := range keyRows {
		pad := strings.Repeat(" ", (i*2)+4)
		var keys []string
		for _, ch := range row {
			v, seen := keyboard[ch]
			upper := unicode.ToUpper(ch)
			if !seen {
				keys = append(keys, fmt.Sprintf(" %c ", upper))
			} else {
				keys = append(keys, fmt.Sprintf("%s %c %s", colorFor(game.LetterResult(v)), upper, reset))
			}
		}
		fmt.Println(pad + strings.Join(keys, ""))
	}
	fmt.Println()
}

// --- Public API ---

func DisplayWelcome() {
	fmt.Println()
	fmt.Printf("    %s W  O  R  D  L  E %s\n", bold, reset)
	fmt.Println()
	fmt.Println("    Guess the 5-letter word in 6 tries.")
	fmt.Println()
	fmt.Printf("    %s E %s correct   ", green, reset)
	fmt.Printf("%s E %s wrong spot   ", yellow, reset)
	fmt.Printf("%s E %s not in word\n", gray, reset)
}

// DisplayTurn redraws the full board and keyboard after every guess.
func DisplayTurn(guesses []game.GuessResult) {
	for _, g := range guesses {
		updateKeyboard(g)
	}
	printBoard(guesses)
	printKeyboard()
}

func ReadGuess(attemptNum int) string {
	fmt.Printf("    Guess %d/%d: ", attemptNum, game.MaxGuesses)
	line, _ := reader.ReadString('\n')
	return strings.TrimSpace(strings.ToLower(line))
}

func DisplayError(msg string) {
	fmt.Printf("    %s-- %s%s\n\n", dim, msg, reset)
}

func DisplayWin(attempts int) {
	msgs := [7]string{"", "Genius!", "Magnificent!", "Impressive!", "Solid!", "Good job!", "Phew!"}
	fmt.Printf("\n    %s%s%s Got it in %d.\n\n", bold, msgs[attempts], reset, attempts)
}

func DisplayLoss(target string) {
	fmt.Printf("\n    The word was %s%s%s. Better luck next time.\n\n",
		bold, strings.ToUpper(target), reset)
}
